### PR TITLE
Enhancement/settings

### PR DESCRIPTION
**Pull request by Geist5000**
Does Add a SettingsContext which makes the settings easaly accessible by all components.

Improved the PaginationLoader Class. 

Added  a UserChanged Event to the Session Class

### CODE REVIEW

**Comment by Geist5000** [comment]
Please do not delete this branch after merging, It could be usefull while merging the refactoring branch

File: src/components/platform/settings/Settings.tsx [hunk 1]
```diff
@@ -2,138 +2,120 @@ import {Component} from "react";
...
 }
 
+/**
+ * Zeigt alle Einstellungen an und gibt die Möglichkeit diese zu ändern
+ */
 export class Settings extends Component<{}, SettingsState> {
 
 
+    /**
+     * Map die ein JSXElement für den angegeben Einstellungstype liefter
```
Comment (nic-schi) [inline src/components/platform/settings/Settings.tsx:42]: *Einstellungstyp liefert

File: src/general-components/API/PaginationLoader.ts [hunk 2]
```diff
@@ -48,15 +54,19 @@ export class PaginationLoader<D extends object> {
         return null;
     }
 
-    public async getAll() {
+    /**
+     * Lädt alle Daten aus dem Backend und potentiel auch aus den Cached und gibt sie als ein Array zurück
```
Comment (nic-schi) [inline src/general-components/API/PaginationLoader.ts:58]: *potenziell oder potentiell

File: src/general-components/API/SettingsCache.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+     */
+    private async loadData(cachedSettings: boolean = false, cachedUserSettings: boolean = false) {
+        // Promise.all damit die Backend calls gleichzeitig stattfinden können und nicht nacheinander passieren müssen
+        const [settings, userSettings] = await Promise.all([this.settingsLoader.getAll(cachedSettings), this.userSettingsLoader.getAll(cachedUserSettings)]);
+        return settings.map((s) => {
+            let userS = userSettings.find((su) => su.setting_id === s.id);
+            let v = s.default;
+            const exists = !!userS && !!userS.value;
+            if (exists) {
+                //TODO make beautiful
```
Comment (nic-schi) [inline src/general-components/API/SettingsCache.ts:56]: this is the best TODO ive ever seen

File: src/general-components/API/SettingsCache.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                v = userS?.value ?? v;
```
Comment (nic-schi) [inline src/general-components/API/SettingsCache.ts:57]: but indeed, this should somehow be made better. maybe move the exists condition into the if or find a workaround.

File: src/general-components/Settings/SettingsList.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    value: string
+    /**
+     * Whether the user already did override the default value
+     */
+    exists: boolean
+}
+
+/**
+ * Klasse um alle Einstellungen verwalten, um Einstellungen hinzuzufügen muss eine neue Instanz erstellt werden
```
Comment (nic-schi) [inline src/general-components/Settings/SettingsList.ts:15]: correct grammar

**Review by nic-schi — CHANGES_REQUESTED** [review]

File: src/general-components/API/SettingsCache.ts (hunk 4 above)
Comment (Geist5000) [reply src/general-components/API/SettingsCache.ts:57]: fixed the logic, now the variable wouldn't make sense anyway

**Pull request merged**